mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/model.scala

```diff
@@ -0,0 +1,1175 @@
+
+  /**
+   * A message has been received that exceeds the maximum we have chosen--which is
+   * typically much less than what the spec permits. The reported sizes do not
+   * include the fixed header size of 2 bytes.
+   */
+  final case class InvalidPacketSize(packetSize: Int, maxPacketSize: Int) extends DecodeError
+
+  /**
+   * Cannot determine the protocol name/level combination of the connect
+   */
+  final case class UnknownConnectProtocol(protocolName: Either[DecodeError, String], protocolLevel: ProtocolLevel)
+      extends DecodeError
+
+  /**
+   * Bit 0 of the connect flag was set - which it should not be as it is reserved.
+   */
+  final case object ConnectFlagReservedSet extends DecodeError
+
+  /**
+   * Something is wrong with the connect message
+   */
+  final case class BadConnectMessage(clientId: Either[MqttCodec.DecodeError, String],
+                                     willTopic: Option[Either[MqttCodec.DecodeError, String]],
+                                     willMessage: Option[Either[MqttCodec.DecodeError, String]],
+                                     username: Option[Either[MqttCodec.DecodeError, String]],
+                                     password: Option[Either[MqttCodec.DecodeError, String]])
+      extends DecodeError
+
+  /**
+   * A reserved QoS was specified
+   */
+  case object InvalidQoS extends DecodeError
+
+  /**
+   * Bits 1 to 7 are set with the Connect Ack flags
+   */
+  case object ConnectAckFlagReservedBitsSet extends DecodeError
+
+  /**
+   * Something is wrong with the publish message
+   */
+  final case class BadPublishMessage(topicName: Either[DecodeError, String],
+                                     packetId: Option[PacketId],
+                                     payload: ByteString)
+      extends DecodeError
+
+  /**
+   * Something is wrong with the subscribe message
+   */
+  final case class BadSubscribeMessage(packetId: PacketId,
+                                       topicFilters: Seq[(Either[DecodeError, String], ControlPacketFlags)])
+      extends DecodeError
+
+  /**
+   * Something is wrong with the subscribe ack message
+   */
+  final case class BadSubAckMessage(packetId: PacketId, returnCodes: Seq[ControlPacketFlags]) extends DecodeError
+
+  /**
+   * Something is wrong with the unsubscribe message
+   */
+  final case class BadUnsubscribeMessage(packetId: PacketId, topicFilters: Seq[Either[DecodeError, String]])
+      extends DecodeError
+
+  /**
+   * JAVA API
+   */
+  final case class DecodeErrorOrControlPacket(v: Either[DecodeError, ControlPacket]) {
+    def getDecodeError: Optional[DecodeError] =
+      v match {
+        case Right(_) => Optional.empty()
+        case Left(de) => Optional.of(de)
+      }
+
+    def getControlPacket: Optional[ControlPacket] =
+      v match {
+        case Right(cp) => Optional.of(cp)
+        case Left(_) => Optional.empty()
+      }
+  }
```

Comment:
@2m I've arrived at this convention for mapping Scala's either to Java. I use it now in a couple of places. It tends to read well when used.